### PR TITLE
[RFC] Provide a `pv moves` command to extract more info from the TT

### PR DESCRIPTION
**Pull request by lakinwecker**
This is a WIP at the moment. Too many TODOs - but I wanted to track it as a PR.

I'll bug @ddugovic on slack/irc tomorrow for answers about what we want to do for all of the TODOs.

### CODE REVIEW

File: src/uci.cpp [hunk 1]
```diff
@@ -169,6 +171,68 @@ namespace {
...
+    if (is >> token && token == "moves") {
+        // Parse move list (if any)
+        while (is >> token && (m = UCI::to_move(pos, token)) != MOVE_NONE)
+        {
+            if (MoveList<LEGAL>(pos).contains(m)) {
+                pvMovesStates->push_back(StateInfo());
+                pos.do_move(m, pvMovesStates->back());
+                moves.push_back(m);
+            } else {
+                //TODO: do what!?
```
Comment (ddugovic) [inline src/uci.cpp]: I think that `position fen FEN moves MOVES` breaks on the first illegal move?

File: src/uci.cpp [hunk 2]
```diff
@@ -169,6 +171,68 @@ namespace {
...
+        /*if (!tb && i == PVIdx)
+            ss << (v >= beta ? " lowerbound" : v <= alpha ? " upperbound" : "");*/
+
+        ss << " pv";
+
+        ss << " " << UCI::move(m, pos.is_chess960());
+        sync_cout << ss.str() << sync_endl;
+    } else {
+        // TODO: what to output here.
+        sync_cout << "no moves" << sync_endl;
```
Comment (ddugovic) [inline src/uci.cpp]: `(none)` is used elsewhere when there is no move

File: src/uci.cpp [hunk 3]
```diff
@@ -169,6 +171,68 @@ namespace {
...
+    if (ttHit) {
+        m = tte->move(); // Local copy to be SMP safe
+
+        Depth d = tte->depth();
+        // TODO: this seems swap between +/- need to figure
+        //       out how to make that consistent, right?
+        Value v = tte->value();
+
+        // TODO: not sure what this does.
+        /*bool tb = TB::RootInTB && abs(v) < VALUE_MATE - MAX_PLY;
```
Comment (ddugovic) [inline src/uci.cpp]: I think if the root position is in the syzygybase, this looks up the win/draw/loss evaluation.

File: src/uci.cpp [hunk 4]
```diff
@@ -169,6 +171,68 @@ namespace {
...
+    TTEntry* tte = TT.probe(pos.key(), ttHit);
+    if (ttHit) {
+        m = tte->move(); // Local copy to be SMP safe
+
+        Depth d = tte->depth();
+        // TODO: this seems swap between +/- need to figure
+        //       out how to make that consistent, right?
+        Value v = tte->value();
+
+        // TODO: not sure what this does.
```
Comment (ddugovic) [inline src/uci.cpp]: I think if the root position is in the syzygybase, this looks up the win/draw/loss evaluation.

File: src/uci.cpp [hunk 5]
```diff
@@ -169,6 +171,68 @@ namespace {
...
+        /*bool tb = TB::RootInTB && abs(v) < VALUE_MATE - MAX_PLY;
+        v = tb ? TB::Score : v;*/
+
+        stringstream ss; 
+        ss << "info"
+           << " depth "    << d / ONE_PLY
+           << " score "    << UCI::value(v);
+
+        // TODO: Not sure how to get access to alpha//beta to calc this.
```
Comment (ddugovic) [inline src/uci.cpp]: Maybe it's accessible from `thread.h`?  I'm honestly not sure.

File: src/uci.cpp [hunk 6]
```diff
@@ -169,6 +171,68 @@ namespace {
...
+        }
+    }
+    bool ttHit = false;
+    TTEntry* tte = TT.probe(pos.key(), ttHit);
+    if (ttHit) {
+        m = tte->move(); // Local copy to be SMP safe
+
+        Depth d = tte->depth();
+        // TODO: this seems swap between +/- need to figure
+        //       out how to make that consistent, right?
```
Comment (ddugovic) [inline src/uci.cpp]: Either that or alpha/beta need to be switched & negated (which might be simpler & easier to interpret).

**Comment by niklasf** [comment]
Let me state for the record that I am not a big fan of introducing a non-standard UCI command to fix ornicar/lila#2423. An approach like #179 looks much more promising to me ...

If there is a good method to decide when to cut off the PV, then Stockfish should just do that, instead of having to implement it in various places (fishnet, ceval, mobile, maybe others). If this should be optional or if it requires parameters how about introducing a new option (via `setoption`) instead?

**Comment by ddugovic** [comment]
Thanks for the feedback and I'll make sure to accept an approach like #179 although @official-stockfish will probably break it as they continue making improvements.

> If there is a good method to decide when to cut off the PV, then Stockfish should just do that...

Well, sure; but there is a trade-off between the PV being accurate and being long, and a month ago I read a Google Forum post indicating that over the years different approaches have been tried.  I'll see if I can invent a new, all-encompassing standard (since maybe such a thing is actually possible).

**Comment by sf-x** [comment]
Stahp!

You _can't_ extract PV from TT if rootMoves[i].pv is short. Historical `extract_pv_from_tt` was _misnamed_. Waht is produced was _not_ a PV. The correct approach is to preserve actual PVs (from all threads if desired), then select one.

**Comment by ddugovic** [comment]
@sf-x I'm curious if you have feedback on the idea of #179 forcing Stockfish to output longer lines (although I think by that same reasoning the longer lines aren't PVs either?)

**Comment by sf-x** [comment]
I think that in current Stockfish, anything longer than 2 plies is should be real PV (but I'm not _completely_ sure).

**Comment by ddugovic** [comment]
Hm, @sf-x is correct - there's a huge inconsistency problem.  For consistent results perhaps `go nodes 1` continuing from the end of the PV is necessary (or returning a stale PV like #179):

```
position startpos
go nodes 100
info depth 1 seldepth 1 multipv 1 score cp 90 nodes 20 nps 10000 tbhits 0 time 2 pv e2e4
info depth 2 seldepth 2 multipv 1 score cp 92 nodes 47 nps 23500 tbhits 0 time 2 pv e2e4 b7b6
info depth 3 seldepth 3 multipv 1 score cp 65 nodes 102 nps 51000 tbhits 0 time 2 pv e2e4 d7d6 d2d3
bestmove e2e4 ponder d7d6

position startpos moves e2e4 d7d6 d2d3
go nodes 1
info depth 1 seldepth 1 multipv 1 score cp -8 nodes 29 nps 9666 tbhits 0 time 3 pv e7e5
info depth 2 seldepth 2 multipv 1 score cp -8 nodes 30 nps 10000 tbhits 0 time 3 pv e7e5
bestmove e7e5
```

**Comment by sf-x** [comment]
>  For consistent results perhaps go nodes 1 continuing from the end of the PV is necessary

Why not produce 50 plies of analysis from 50 depth-1 searches this way. The users will be amazed by the speed and depth! :P

> (or returning a stale PV like #179 in which case perhaps we should also use the stale evaluation

The only sane solution, of course. They are not necessarily "stale": rootMoves[i].pv is not guaranteed to contain an actual PV (except just after a search iteration finishes).

**Comment by ddugovic** [comment]
> Why not produce 50 plies of analysis from 50 depth-1 searches this way. The users will be amazed by the speed and depth! :P

Indeed... thanks for correcting my woefully mistaken understanding of `extract_pv_from_tt` and sorry to @lakinwecker that I led us astray - we can't rely on the TT after all.  Having a `pv` command might be useful for repeating and/or filtering score/PV data in #179 if we have access to the threads.

Speaking of speed, I'm curious: supposing I have one minute in which to annotate a 30-move game, is it a bad idea to spend 1 second per position?

**Comment by sf-x** [comment]
> Speaking of speed, I'm curious: supposing I have one minute in which to annotate a 30-move game, is it a bad idea to spend 1 second per position?

If you use multiple cores, I don't know of anything much better (but see https://github.com/official-stockfish/Stockfish/pull/725 and below).
If you use one core, that's a bad idea: you should try to complete an iteration. The simplest way to do that is to use Stockfish's native time allocation (give `go wtime X winc Y` for some X and Y instead of `go movetime X`. This will abort the search if there's low chance to complete an iteration (if time used to complete the last iterations is more than Z% of optimum, IIRC).

**Comment by ddugovic** [comment]
Interesting, I'll review the iteration time allocation and see I can derive a "reverse-iteration" time allocation function (for postmortem analysis starting from the final position and working backwards to the first position).

**Comment by lakinwecker** [comment]
Hi All,

Thanks for the information. All great stuff.  I wondered about the TT values and whether they would be useful or not.  I guess we know now. 

Here is what confuses me about what you have said.  If I put a new position into a just launched stockfish and ask it to `go movetime 4000 nodes 3500000`, it will sometimes return a lengthy PV, sometimes a PV of length 1. When it returns the PV of length 1 there is plenty of output with lengthy PVs before that point. But sometimes when we get to the end, all of the threads have a `rootMoves[0].pv` that is 1 length.  But in this isolated case, I don't understand?  They would obviously have had to have calculated further than one ply to determine the `bestMove`

What confuses me is that the TT is empty when we start, and other threads will report lengthy PVs.  But why would it _ever_ have a short PV in that case? Do threads calculate a PV, then store the root move in the TT and forget the rest and move on to something that's turns out to be worse?

Would a sane approach be to modify the TTEntry to have a PV and have the threads set that whenever they set the TTEntry?

**Comment by ddugovic** [comment]
In theory I can understand that because a search was terminated early (without finding mate and without exhausting the search space) the [principal variation](http://chessprogramming.wikispaces.com/Principal+variation) may be short as Stockfish may be confused about what moves constitute best play.  Often OTB I am forced to make moves with an incomplete understanding.

In the case of 1-move PVs, maybe what occurred is that the old PV was partially or completely refuted on move 2, or that another variation was found to be equally good and it is unclear which variation is best?

The TT is just that - an aging directed graph for remembering evaluations and moves.  AFAIK it does not automatically update all dependencies in the graph whenever any value changes; so if two variations reach the same position (at increasing depths such that the second variation searches and updates the TT) I don't think all the nodes in the first variation automatically update, even when the first variation is the PV!

**Comment by sf-x** [comment]
> Here is what confuses me about what you have said. If I put a new position into a just launched stockfish and ask it to go movetime 4000 nodes 3500000, it will sometimes return a lengthy PV, sometimes a PV of length 1. When it returns the PV of length 1 there is plenty of output with lengthy PVs before that point. But sometimes when we get to the end, all of the threads have a rootMoves[0].pv that is 1 length. But in this isolated case, I don't understand? They would obviously have had to have calculated further than one ply to determine the bestMove

**`rootMoves[i].pv` is not guaranteed to contain an actual PV (except just after a search iteration finishes).**

**Comment by sf-x** [comment]
In case of a fail-high or a fail-low, there's no PV.

**Comment by lakinwecker** [comment]
Thanks you've clarified it as much as is possible without me reviewing the search algorithm in depth. 

I'm closing this. This approach is not going to work.

**Pull request closed without merging**